apps/pipelines/models.py

```diff
@@ -116,8 +117,17 @@ def update_nodes_from_data(self) -> None:
         # Delete old nodes
         current_ids = set(self.node_ids)
         new_ids = set(node.id for node in nodes)
-        to_delete = current_ids - new_ids
-        Node.objects.filter(pipeline=self, flow_id__in=to_delete).delete()
+        to_remove = current_ids - new_ids
+
+        pipeline_nodes = Node.objects.annotate(versions_count=models.Count("versions")).filter(
+            pipeline=self, flow_id__in=to_remove
+        )
+        nodes_to_archive = pipeline_nodes.filter(versions_count__gt=0)
+        pipeline_nodes.filter(versions_count=0).delete()
+
+        for node in nodes_to_archive:
+            # Preserve the node if it has versions, otherwise we tamper with previous versions
+            node.archive()
```

Comment:
I think one thing we need to iron out still with versioning is when to delete-- I think originally the workflow was active -> archived -> deleted, but this would make it such that a node would not be able to be deleted once archived. I think its out of scope for this PR since it's already reaching was past just a diff view but working discussing